packages/core/src/browser/quick-open/quick-command-service.ts

```diff
@@ -109,15 +113,22 @@ export class QuickCommandService implements QuickOpenModel, QuickOpenHandler {
         // Get the list of all valid commands.
         const allCommands: Command[] = this.getValidCommands(this.commands.commands);
 
+        // Get the max history limit.
+        const limit: number = this.corePreferences['workbench.commandPalette.history'];
+
         // Build the list of recent commands.
         const rCommands: Command[] = [];
         recentCommands.forEach((r: Command) => {
+            // Opt out of displaying the recently used list.
+            if (limit === 0) {
+                return;
+            }
             // Determine if the command is exempted from display.
             const exempted: boolean = this.exemptedCommands.some((c: Command) => Command.equals(r, c));
             // Determine if the command currently exists in the list of all available commands.
             const exists: boolean = allCommands.some((c: Command) => Command.equals(r, c));
             // Add the recently used item to the list.
-            if (exists && !exempted) {
+            if (exists && !exempted && rCommands.length < limit) {
```

Comment:
Should not we always insert most recent at the top and remove oldest?

Reply:
It is already handled in `commands.ts`

https://github.com/theia-ide/theia/blob/aff18c8f31661e18b22d05f71b56b4ed65975093/packages/core/src/common/command.ts#L366-L377